### PR TITLE
BLE connectivity

### PR DESCRIPTION
**Pull request by t-pi**
Okay, here's a first BT BLE PR. I've marked it as draft at first, although it runs on both my test devices, on Lorastick with BME680 and one Wifi Stick 32. Did some further minor changes, e.g.
- rename "display" to "publish", as it also sends to BT
- if new status change, push directly to display

### CODE REVIEW

**Comment by t-pi** [comment]
..on the other hand, it works on both chips, this is a -dev version. Removed "draft"

**Comment by t-pi** [comment]
For checking, I have made good experiences as well on iOS as on Android with nordic's "nRF Connect" app to connect via BLE to Multigeiger, read characteristics and start notifications from "heart rate measurement" service. These should correspond to the main displayed CPM value with a minimal delay.

File: multigeiger/multigeiger.ino [hunk 1]
```diff
@@ -65,10 +66,19 @@ void setup() {
   setup_display(isLoraBoard);
   setup_switches(isLoraBoard);
   switches = read_switches();  // only read DIP switches once at boot time
+  if (!sendToBle)
+    switches.ble_on = false; // avoid switching on in WebConfig and trying to use BLE without setup
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino:70]: that's a bit dirty, the switches.xxx should just reflect the hw switches.

File: multigeiger/multigeiger.ino [hunk 2]
```diff
@@ -65,10 +66,19 @@ void setup() {
...
   setup_webconf(isLoraBoard);
   setup_speaker(playSound, ledTick && switches.led_on, speakerTick && switches.speaker_on);
   setup_transmission(VERSION_STR, ssid, isLoraBoard);
+  // BT BLE
+  if (sendToBle && switches.ble_on) {
+    set_status(STATUS_BT, ST_BT_INIT);
+    setup_ble(ssid);
+  } else {
+    set_status(STATUS_BT, ST_BT_OFF);
+  }
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino:81]: just move this code **into** setup_ble() so it is just calling that from here?

File: multigeiger/multigeiger.ino [hunk 3]
```diff
@@ -104,7 +114,21 @@ int update_wifi_status(void) {
   return st;
 }
 
-void display(unsigned long current_ms, unsigned long current_counts, unsigned long gm_count_timestamp, unsigned long current_hv_pulses) {
+int update_bt_status(void) { // currently no error detection
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino:117]: try to always have 2 blanks before //.

File: multigeiger/multigeiger.ino [hunk 4]
```diff
@@ -104,7 +114,21 @@ int update_wifi_status(void) {
...
+  int st;
+  if (sendToBle && switches.ble_on) {
+    if (is_ble_connected())
+      st = ST_BT_CONNECTED;
+    else
+      st = ST_BT_CONNECTABLE;
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino:123]: could be another `?:`.

File: multigeiger/multigeiger.ino [hunk 5]
```diff
@@ -260,8 +288,9 @@ void loop() {
   set_status(STATUS_HV, hv_error ? ST_HV_ERROR : ST_HV_OK);
 
   int wifi_status = update_wifi_status();
+  update_bt_status();
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino:291]: looks like we have some `bt` and some `ble` now.

File: multigeiger/userdefines-example.h [hunk 6]
```diff
@@ -56,3 +56,10 @@
 // Note: The TTN configuration needs to be done in lorawan.cpp (starting at line 65).
 #define SEND2LORA false
 
+// Send data via BLE?
+// Device provides "Heart Rate Service" (0x180D) and these characteristics
```
Comment (ThomasWaldmann) [inline multigeiger/userdefines-example.h:60]: Hmm, is there nothing rather generic we could use so we must abuse a `Heart Rate` measuring mechanism?

Comment (ThomasWaldmann) [reply multigeiger/userdefines-example.h:60]: OTOH, if the CPM goes rather high, your heart rate might also go up. :-X

File: multigeiger/ble.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+#include "userdefines.h"
+#include "ble.h"
+
+#include <BLEDevice.h>
+#include <BLEServer.h>
+#include <BLEUtils.h>
+#include <BLE2902.h>
+
+#define BLE_SERVICE_HEART_RATE    BLEUUID((uint16_t)0x180D)  // 16bit UUID of Heart Rate Service
+#define BLE_CHAR_HR_MEASUREMENT   BLEUUID((uint16_t)0x2A37)  // 16 bit UUID of Heart Rate Measurement Characteristic
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:24]: either have the blank after 16 or not, be consistent.

File: multigeiger/display.cpp [hunk 8]
```diff
@@ -35,18 +35,16 @@ void display_start_screen(void) {
...
+    pu8x8->drawString(0, 4, line);
   } else {
     pu8x8->setFont(u8x8_font_amstrad_cpc_extended_f);
-    pu8x8->drawString(0, 0, "Geiger-Counter");
-    pu8x8->drawString(0, 2, "==============");
+    pu8x8->drawString(0, 0, "Multi-Geiger");
+    pu8x8->drawString(0, 2, "Info:boehri.de");
     snprintf(line, 15, "%s", VERSION_STR);  // 14 chars + \0 termination
     pu8x8->drawString(0, 4, line);
-    pu8x8->drawString(0, 6, "Info:boehri.de");
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/display.cpp:49]: Right. I am preparing a separate PR for the display changes. Due to the instant status line update upon changes, the last line got overwritten too fast...

File: multigeiger/display.cpp [hunk 9]
```diff
@@ -91,17 +89,20 @@ static const char *status_chars[STATUS_MAX] = {
...
-    status[index] = value;
-  else
+  if ((index >= 0) && (index < STATUS_MAX)) {
+    if (status[index] != value) {
+      status[index] = value;
+      displayStatus();
+    }
+  } else
     log(ERROR, "invalid parameters: set_status(%d, %d)", index, value);
 }
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/display.cpp:107]: will reflect in separate PR

File: multigeiger/ble.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+bool is_ble_connected() {
+  return deviceConnected;
+}
+
+class MyServerCallbacks: public BLEServerCallbacks {
+  void onConnect(BLEServer *pServer, esp_ble_gatts_cb_param_t *param) {
+    char remoteAddress[18];
+    sprintf(
+      remoteAddress,
+      "%.2X:%.2X:%.2X:%.2X:%.2X:%.2X",
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:54]: integrated this code during debugging, will check with utils.cpp

File: multigeiger/ble.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+      status_HRCP = 0;
+    }
+  }
+};
+
+void update_bledata(unsigned int cpm) {
+  cpm_update_counter++;
+  cpm_update_counter = cpm_update_counter & 0xFFFF;
+  if (status_HRCP > 0) {
+    log(INFO, "HR Control Point Value received: %d, resetting packet counter", status_HRCP);
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:88]: agreed

File: multigeiger/ble.cpp [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  txBuffer_HRM[3] = (cpm_update_counter & 0xFF);
+  txBuffer_HRM[4] = (cpm_update_counter >> 8) & 0xFF;
+  bleCharHRM.setValue(txBuffer_HRM, 5);
+  bleCharHRM.notify();
+}
+
+void setup_ble(char *device_name) {
+  BLEDevice::init(device_name);
+  // Create the BLE Server
+  BLEServer *bleServer = BLEDevice::createServer();
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:105]: okay, will remove these unecessary comments and align in-line comments

File: multigeiger/ble.cpp [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+  bleCharHRM.addDescriptor(&bleDescriptorHRM);
+  bleCharHRM.addDescriptor(new BLE2902());  // for internal notification management
+
+  bleService->addCharacteristic(&bleCharHRCP);
+  bleDescriptorHRCP.setValue("0x01 for count reset");
+  bleCharHRCP.addDescriptor(&bleDescriptorHRCP);
+  bleCharHRCP.setCallbacks(new MyCharacteristicCallbacks());
+
+  bleService->addCharacteristic(&bleCharHRPOS);
+  bleDescriptorHRPOS.setValue("Geiger Mueller Tube Type 0-3");
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:123]: yes, noticed this from your discussion already, too...

File: multigeiger/multigeiger.ino (hunk 1 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino:70]: Hmm, any other idea how to prevent "using BLE" if not initialized? Another status variable will make the code uglier at many more calls.
My interpretation was, that we need indeed some "HW switch"...

File: multigeiger/multigeiger.ino (hunk 2 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino:81]: I wanted to avoid mixing too many modules, but I guess a status call from Lora and BLE is completely understandable. Will move it there.

File: multigeiger/multigeiger.ino (hunk 5 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino:291]: yes, indeed, I always used ble in my code, but saw your bt in the main .ino and tried to stay consistent - but wasn't of course... Would it be okay to change everything to ble? We are using Bluetooth Low Energy, and 3 letters hint more than 2.

File: multigeiger/userdefines-example.h (hunk 6 above)
Comment (t-pi) [reply multigeiger/userdefines-example.h:60]: The intention was that this way the Multigeiger can be connected to different devices to display the CPM - all that can display Heart Rate via BLE. But typically not both CPM and real Heart Rate, though :)

**Comment by t-pi** [comment]
Need other PR for display changes first...

File: multigeiger/multigeiger.ino (hunk 4 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino]: done

File: multigeiger/multigeiger.ino (hunk 3 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino]: ok

File: multigeiger/ble.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp]: hmm, found no easy way to generalize this with my limited C++ knowledge. Did not want to risk some memory leaks... 
Left it as it is so far.

[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp]: ...except that indeed the . is now a 0

[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/ble.cpp]: ah, right, of course it must not be %2X, but %02X.

File: multigeiger/multigeiger.ino (hunk 5 above)
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: Guess both bt and ble is fine. One is more general, one more specific.

File: multigeiger/ble.cpp [hunk 14]
```diff
@@ -0,0 +1,141 @@
+// Project: Simple Multi-Geiger
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:1]: clear anyway - remove?

File: multigeiger/ble.cpp [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+// Modul: BLE transmission of CPM data
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:2]: "Module" (or remove).

also I would not specify what data is transferred, that can change any time and you don't want to have to update that line all the time.

File: multigeiger/ble.cpp [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+//
+// code related to the transmission of the measured data via Bluetooth Low Energy
+// Uses GATT Heart Rate Measurement Service for notifications with CPM update
+//
+// Heart Rate Measurement = Radiation CPM, Energy Expense = Rolling Packet Counter
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:7]: the heart rate part is somehow duplicate.

File: multigeiger/display.cpp [hunk 17]
```diff
@@ -35,18 +35,16 @@ void display_start_screen(void) {
...
+    pu8x8->drawString(0, 4, line);
   } else {
     pu8x8->setFont(u8x8_font_amstrad_cpc_extended_f);
-    pu8x8->drawString(0, 0, "Geiger-Counter");
-    pu8x8->drawString(0, 2, "==============");
+    pu8x8->drawString(0, 0, "Multi-Geiger");
+    pu8x8->drawString(0, 2, "Info:boehri.de");
     snprintf(line, 15, "%s", VERSION_STR);  // 14 chars + \0 termination
     pu8x8->drawString(0, 4, line);
-    pu8x8->drawString(0, 6, "Info:boehri.de");
```
Comment (ThomasWaldmann) [inline multigeiger/display.cpp:49]: changes are still in here?

File: multigeiger/display.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/display.cpp:107]: that other PR looks like we could soon merge it. you also need to remove it from here.

File: multigeiger/userdefines-example.h (hunk 6 above)
Comment (ThomasWaldmann) [reply multigeiger/userdefines-example.h:60]: ah, ok.

**Comment by rexfue** [comment]
@t-pi: how to test it on my phone? Which app can i use?

**Comment by t-pi** [comment]
@rexfue I have made good experiences with nordic nRF Connect app. You have to connect and then start notifications to get the data (click the corresponding icon of heart rate measurement characteristic)

Will try to set up repo for flutter app over the long weekend. It was only tested on Android so far, though...

**Comment by rexfue** [comment]
OK, thanks. Is it somehow important, that the values for cpm and counter are sent in little endian?

**Comment by t-pi** [comment]
Well, that's the way if we want to be compatible to the heart rate service (at least the implementation from nordic :)

File: multigeiger/ble.cpp (hunk 14 above)
Comment (t-pi) [reply multigeiger/ble.cpp:1]: ok

File: multigeiger/ble.cpp (hunk 15 above)
Comment (t-pi) [reply multigeiger/ble.cpp:2]: ok, will change to Multigeiger data

File: multigeiger/ble.cpp (hunk 16 above)
Comment (t-pi) [reply multigeiger/ble.cpp:7]: yeah, I kept the official names...

File: multigeiger/display.cpp (hunk 17 above)
Comment (t-pi) [reply multigeiger/display.cpp:49]: some changes are required for the PR to run, will remove, rebase and squash as soon as the other PR is through.

**Comment by ThomasWaldmann** [comment]
@t-pi please have a look at the other PR and if its contents look ok to you, tell me, then i'll rebase this one onto the other.

**Comment by ThomasWaldmann** [comment]
rebased on master, cleaned up. please review and also practically test.

File: multigeiger/multigeiger.ino (hunk 5 above)
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: to resolve this: make a ticket and a separate PR. it is a cosmetic cleanup, should not be in here.

Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: #315

File: multigeiger/ble.cpp [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+  if (!ble_enabled)
+    return;
+  cpm_update_counter++;
+  cpm_update_counter = cpm_update_counter & 0xFFFF;
+  if (status_HRCP > 0) {
+    log(DEBUG, "HR Control Point Value received: %d, resetting packet counter", status_HRCP);
+    cpm_update_counter = 0;
+    status_HRCP = 0;
+  }
+  cpm_value = cpm;
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:90]: why is that line needed? why is the global `cpm_value` needed?

File: multigeiger/ble.cpp [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+
+  BLEServer *bleServer = BLEDevice::createServer();
+  bleServer->setCallbacks(new MyServerCallbacks());
+
+  BLEService *bleService = bleServer->createService(BLE_SERVICE_HEART_RATE);
+
+  bleService->addCharacteristic(&bleCharHRM);
+  bleDescriptorHRM.setValue("Radiation rate CPM");
+  bleCharHRM.addDescriptor(&bleDescriptorHRM);
+  bleCharHRM.addDescriptor(new BLE2902());  // required for internal notification management
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:118]: "internal" refers to what?

File: multigeiger/ble.cpp [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+    log(DEBUG, "HR Control Point Value received: %d, resetting packet counter", status_HRCP);
+    cpm_update_counter = 0;
+    status_HRCP = 0;
+  }
+  cpm_value = cpm;
+  txBuffer_HRM[0] = flags_HRS;
+  txBuffer_HRM[1] = cpm_value & 0xFF;
+  txBuffer_HRM[2] = (cpm_value >> 8) & 0xFF;
+  txBuffer_HRM[3] = cpm_update_counter & 0xFF;
+  txBuffer_HRM[4] = (cpm_update_counter >> 8) & 0xFF;
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:95]: for what is the update counter needed?

File: multigeiger/ble.cpp [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+
+  BLEService *bleService = bleServer->createService(BLE_SERVICE_HEART_RATE);
+
+  bleService->addCharacteristic(&bleCharHRM);
+  bleDescriptorHRM.setValue("Radiation rate CPM");
+  bleCharHRM.addDescriptor(&bleDescriptorHRM);
+  bleCharHRM.addDescriptor(new BLE2902());  // required for internal notification management
+
+  bleService->addCharacteristic(&bleCharHRCP);
+  bleDescriptorHRCP.setValue("0x01 for count reset");
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:121]:
```
... for update counter reset
```

just to not confuse it with GM counts.

File: multigeiger/ble.h [hunk 22]
```diff
@@ -0,0 +1,11 @@
+// Code related to the transmission of the measured data via Bluetooth Low Energy
+// Uses GATT Heart Rate Measurement Service for notifications with CPM update
```
Comment (ThomasWaldmann) [inline multigeiger/ble.h:2]: see above.

File: multigeiger/ble.h [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+
+#ifndef _BLE_H_
+#define _BLE_H_
+
+void setup_ble(char *device_name, bool ble_enabled);
+void update_bledata(unsigned int cpm);
+bool is_ble_connected();
```
Comment (ThomasWaldmann) [inline multigeiger/ble.h:9]: (void)

File: multigeiger/multigeiger.ino [hunk 24]
```diff
@@ -104,7 +106,16 @@ int update_wifi_status(void) {
...
 
-void display(unsigned long current_ms, unsigned long current_counts, unsigned long gm_count_timestamp, unsigned long current_hv_pulses) {
+int update_ble_status(void) {  // currently no error detection
+  int st = ST_BT_OFF;
+  if (get_status(STATUS_BT) != ST_BT_OFF) {
+    st = is_ble_connected() ? ST_BT_CONNECTED : ST_BT_CONNECTABLE;
+    set_status(STATUS_BT, st);
+  }
+  return st;
+}
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino:116]: The reverse reading via get_status() is a bit strange.

Couldn't we do it just like:
```
int st;
if (sendToBle)
  st = is_ble_connected() ? ST_BT_CONNECTED : ST_BT_CONNECTABLE;
else
  st = ST_BT_OFF;
set_status(STATUS_BT, st);
return st;
```

File: multigeiger/multigeiger.ino [hunk 25]
```diff
@@ -69,6 +70,7 @@ void setup() {
   setup_webconf(isLoraBoard);
   setup_speaker(playSound, ledTick && switches.led_on, speakerTick && switches.speaker_on);
   setup_transmission(VERSION_STR, ssid, isLoraBoard);
+  setup_ble(ssid, sendToBle);
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino:73]: should ` && switches.ble_on` be added here?

File: multigeiger/ble.cpp [hunk 26]
```diff
@@ -0,0 +1,138 @@
+// Code related to the transmission of the measured data via Bluetooth Low Energy
+// Uses GATT Heart Rate Measurement Service for notifications with CPM update
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/ble.cpp:2]: (upper/lower case, punctuation)

File: multigeiger/ble.cpp [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+
+uint8_t flags_HRS = 0b00001001; // bit 0 --> 1 = HR (cpm) as UINT16, bit 3 --> add UINT16 Energy Expended (rolling update counter)
+uint8_t txBuffer_HRM[5]; // transmit buffer, byte[0] = flags, [1, 2] = cpm, [3, 4] = rolling update counter
+uint8_t txBuffer_HRPOS[1] = {TUBE_TYPE};
+unsigned int status_HRCP = 0;
+unsigned int cpm_update_counter = 0;
+unsigned int cpm_value = 0;
+
+bool is_ble_connected() {
+  return ble_enabled && device_connected;
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:46]: This way, if ble is switched off mid-operation via web interface and ble_enabled set accordingly, it will at least stop notifying and indicate correct status.

File: multigeiger/ble.cpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:2]: ok

File: multigeiger/ble.cpp [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+  void onDisconnect(BLEServer *pServer) {
+    device_connected = false;
+    log(INFO, "BLE device disconnected");
+  }
+};
+
+// Callback allowing for Control Point Characteristic to reset "energy expenditure" (packet counter)
+class MyCharacteristicCallbacks: public BLECharacteristicCallbacks {
+  void onWrite(BLECharacteristic *pCharacteristic) {
+    std::string rxValue = pCharacteristic->getValue();
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:75]: The espressif libs use std::string anyway, however I found I can use .cstr()
Ref: https://arduino.stackexchange.com/questions/56365/arduino-convert-stdstring-to-string

File: multigeiger/ble.cpp [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+    status_HRCP = (rxValue.length() > 0) ? rxValue[0] : 0;
```
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:76]: Good point. Casting to (unsigned int)

File: multigeiger/ble.cpp (hunk 19 above)
Comment (t-pi) [reply multigeiger/ble.cpp:118]: Service internal. Rephrased for better understanding.

File: multigeiger/ble.cpp (hunk 20 above)
Comment (t-pi) [reply multigeiger/ble.cpp:95]: To verify that we have indeed an new value for CPM resp. to account for missed packets. I'm currently using it mainly for some coarse uptime estimation on app side, too, that's why it's 16bit and not only 8bit...

File: multigeiger/ble.cpp (hunk 21 above)
Comment (t-pi) [reply multigeiger/ble.cpp:121]: right, rephrased.

File: multigeiger/multigeiger.ino (hunk 24 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino:116]: Right. I wanted to avoid a status line update every time we run the main loop, but the set_status() is indeed smart enough to update only when actually changed. Will revert here.
Another PR can then indeed remove the get_status() function from display.cpp, as it was only needed here...

File: multigeiger/multigeiger.ino (hunk 25 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino:73]: done

File: multigeiger/ble.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:45]: So far, though, no call to set ble_enabled to false after switch off is implemented. Where would that call go ideally?

File: multigeiger/ble.cpp (hunk 18 above)
Comment (t-pi) [reply multigeiger/ble.cpp]: Currently it is not needed. Will remove it

File: multigeiger/display.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/display.cpp]: ok, done in other PR

File: multigeiger/multigeiger.ino (hunk 1 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino]: Resolved by BLE module's inner state variable

File: multigeiger/multigeiger.ino (hunk 2 above)
Comment (t-pi) [reply multigeiger/multigeiger.ino]: done

**Comment by t-pi** [comment]
switched off draft, as prior display PR is merged.

File: multigeiger/ble.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/ble.cpp:45]: not sure what the protocol is for disconnecting.

but i agree, if the web interface is used to switch off ble_enabled, we need to react somehow. that reaction might include setting `device_connected = false` though.

**Comment by ThomasWaldmann** [comment]
I suddenly see old commits again, what happened?

**Comment by ThomasWaldmann** [comment]
maybe this kind of workflow needs explaining.

as i did major cleanup / rebasing / conflict resolution work, i did not just add more commits to what was on github, but replaced all commits in this PR by a single clean one, using `push -f`.

the branch in your (@t-pi) local git still had the old stuff and was out of sync with your github repo due to my action.

guess the easiest way would have been to kill your local branch and fetch it fresh from github and check it out. only after that start to add more work.

i still have a clean single-commit state in my local repo, so I can push -f again if you like, erasing everything that is here currently.

File: multigeiger/ble.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:45]: The device_connected is set by the callback in the lib. We can call now setup_ble("", false) to set ble_enabled to false. But where should we set this? Do we need a check in the mainloop for the global parameter to detect its change from the web interface?

**Comment by t-pi** [comment]
Hmm, my (limited) understanding was, that the ble_connectivity branch exists only in my t-pi fork and is merged from there to the master branch here. Thus, deleting the fork-branch would delete also this PR here - but apparently that's not it. 
Where would I get the clean ble_connectivity branch?

**Comment by t-pi** [comment]
I did a local rebase to ecocurious' master (local command: git pull --rebase upstream master) and then a push to t-pi/fork's master. Then a (git rebase master) of my local branch "ble_connectivity". Where did it got the old commits from?

**Comment by ThomasWaldmann** [comment]
I guess you still had them locally and you did not remove them.

As I said, I still have the single clean commit in my local repo and I can push -f that to here.
Just make sure you do not lose work if I do that.

**Comment by t-pi** [comment]
OK, go. What should I do afterwards? Rebase my local branch "ble_connectivity" to upstream?

**Comment by ThomasWaldmann** [comment]
Get rid of your local branch first (export whatever you want to keep).

Fetch the branch from github, after I pushed.

Checkout the branch and apply what's missing.

**Comment by t-pi** [comment]
okay, seems to have worked this time. I had to fetch the specific commit from above...

File: multigeiger/ble.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (t-pi) [reply multigeiger/ble.cpp:45]: Also, it was a bit dirty, using setup_ble(), so I've added a specific function disable_ble() to disable ble (= set ble_enable to false) after config change. Question remains - is there a callback after config was changed?

File: multigeiger/ble.cpp [hunk 30]
```diff
@@ -0,0 +1,141 @@
+// Code related to the transmission of the measured data via Bluetooth Low Energy,
+// uses GATT Heart Rate Measurement Service for notifications with CPM update
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:2]: add the `.` at the end of the sentence, please.

File: multigeiger/ble.cpp [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+  bleServer->getAdvertising()->start();
+
+  set_status(STATUS_BT, ST_BT_CONNECTABLE);
+  log(INFO, "BLE service advertising started, device name: %s, MAC: %s", device_name, BLEDevice::getAddress().toString().c_str());
+}
+
+void disable_ble(void) {
+  ble_enabled = false;
+  set_status(STATUS_BT, ST_BT_OFF);
+}
```
Comment (ThomasWaldmann) [inline multigeiger/ble.cpp:141]: add a linefeed, so diff does not complain.

File: multigeiger/ble.h [hunk 32]
```diff
@@ -0,0 +1,12 @@
+// Code related to the transmission of the measured data via Bluetooth Low Energy,
+// uses GATT Heart Rate Measurement Service for notifications with CPM update
```
Comment (ThomasWaldmann) [inline multigeiger/ble.h:2]: add a `.`

**Comment by t-pi** [comment]
All done.

**Comment by t-pi** [comment]
One more thing: in webconf.cpp the CONFIG_VERSION is defined:
```
#define CONFIG_VERSION "013"  // for IoTWebConfig
```
Should we update this to 015, with the changes for the BLE switch?

**Comment by rexfue** [comment]
We can, but I wouldn't recommend it. If config value is changed, ALL values have to be entered again, even wifi credentials.
If we don't change the layout of the configs, only append some new values, we don't need to change the config version (this is, what I have figured out). I will add a comment.

**Comment by ThomasWaldmann** [comment]
@rexfue merge or do you want to review/test again?

**Comment by rexfue** [comment]
merge

**Comment by ThomasWaldmann** [comment]
\o/

**Comment by t-pi** [comment]
Thanks a lot for all the guidance 😄
